### ISSUE TITLE
ambiguity on reaction to RST_STREAM

### ISSUE THREAD

**Issue by mirjak** [quicwg/base-drafts#1671]
The transport draft leaves it currently open to the receiver to delivery outstanding and newly received data to the application after RST_STREAM was received. This introduces some ambiguity for flow control (see section "Edge Caeses and Other Considerations"). Further it may incentive wrong use of RST_STREAM and provide unepxected behavior between different quic implementation ("I always used RST here and data was correctly delivered...").

I would recommend to clearly specify that all data must be immediately dropped if RST_STREAM is received because that should only be send in an error case where that data is not useful anymore anyway.

(Effectly you don't even need the final offset in the RST_STREAM frame in that case anymore, but I guess that doesn't hurt.)

**Comment by mikkelfj** [comment]
You need the offset to account for flow control credits against global data limits. If the stream always ends at a given offset, RST or FIN, how is that ambigous?

**Comment by mirjak** [comment]
If you drop all data on receive of RST of that stream, you don't need to count any of that data for flow control. 

It's ambiguous because currently the receiver can decide to drop and not count it, or to keep it and count, so both ends may be out of sync, as explained on the "Edge Caeses and Other Considerations" section.

**Comment by mikkelfj** [comment]
OK, maybe I'm not up to speed on the spec then.

But as I see it, anything you send on a stream counts as send whether you later decide to RST or FIN. You have used capacity on the wire. The fact that you do not care about the data any more is just a sunk cost, just as if you bought a concert ticket and got sick.

**Comment by mirjak** [comment]
Actually, sorry you are right, you still need the final offset.

However, what this is ambiguous is if data gets delivered to the application or not.

**Comment by mikkelfj** [comment]
Yes, delivery is ambigous but that is not necessarily a bad thing. Ordered delivery isn't even required so you could have deliver some post gap data before seeing a RST - for some application that is a feature. This behaviour is better specified at the application protocol level, but I think a recommended default is in place. For deliver, the recommended behavior is to be in order, by default, I believe.
